File: python/简单工厂模式.py
# coding = utf-8

# 抽象产品类
class Dog():
    @staticmethod
    def sound():
        print("wang wang")

    def introduce(self):
        raise NotImplemented

# 具体产品类1
class WangCai(Dog):
    def introduce(self):
        print("my name is WangCai")
        self.sound()

# 具体产品类2
class AHuang(Dog):
    def introduce(self):
        print("my name is AHuang")
        self.sound()

# 工厂类
class DogStore():
    def __init__(self):
        self.dog = Dog()

    def choice(self, name):
        if name == "WangCai":
            self.dog = WangCai()
        elif name == "AHuang":
            self.dog = AHuang()
        return self.dog


if __name__ == '__main__':
    dog = DogStore()
    dog.choice("WangCai").introduce()
    print("=====================")
    dog.choice("AHuang").introduce()

File: C++/简单工厂模式.cpp
#include<iostream>
#include<string>

using namespace std;

// 抽象产品类Dog 
class Dog
{
	public:
		void sound(){ cout << "wang wang" << endl; }
		virtual void introduce()=0;
	
};

// 具体产品类：小狗旺财 
class RollDog : public Dog
{
	void introduce()
	{
		cout << "my ear is roll ";
		sound(); 
	}
};

// 具体产品类： 小狗阿黄 
class StraightDog : public Dog
{
	void introduce()
	{
		cout << "my ear is straight ";
		sound();
	}
	
};

// 工厂类 
class DogStore
{
	public:
	Dog* dog_create(string name)
	{
		if(name == "RollDog")
			return new RollDog();
		else if(name == "StraightDog")
			return new StraightDog();
	}
	
};


int main()
{
	DogStore factory;
	Dog *rd = factory.dog_create("RollDog");
	rd->introduce();
	Dog *sd = factory.dog_create("StraightDog");
	sd->introduce();
	delete rd;
	delete sd;
	 
	return 0;
 } 
